lib/SILOptimizer/Utils/OwnershipOptUtils.cpp

```diff
@@ -684,65 +737,244 @@ SILBasicBlock::iterator OwnershipRAUWUtility::perform() {
-  if (m->getStage() == SILStage::Raw)
-    return false;
+OwnershipRAUWHelper::OwnershipRAUWHelper(OwnershipFixupContext &inputCtx,
+                                         SingleValueInstruction *inputOldValue,
+                                         SILValue inputNewValue)
+    : ctx(&inputCtx), oldValue(inputOldValue), newValue(inputNewValue) {
+  // If we are already not valid, just bail.
+  if (!isValid())
+    return;
 
-  // If our old ownership kind is ValueOwnershipKind::None and our new kind is
-  // not, we may need to do more work that has not been implemented yet. So
-  // bail.
+  // Otherwise, lets check if we can perform this RAUW operation. If we can't,
+  // set ctx to nullptr to invalidate the helper and return.
+  if (!canFixUpOwnershipForRAUW(oldValue, newValue)) {
+    ctx = nullptr;
+    return;
+  }
+
+  // If we have an object, at this point we are good to go so we can just
+  // return.
+  if (newValue->getType().isObject())
+    return;
+
+  // But if we have an address, we need to check if new value is from an
+  // interior pointer or not in a way that the pass understands. What we do is:
   //
-  // Due to our requirement that types line up, this can only occur given a
-  // non-trivial typed value with None ownership. This can only happen when
-  // oldValue is a trivial payloaded or no-payload non-trivially typed
-  // enum. That doesn't occur that often so we just bail on it today until we
-  // implement this functionality.
-  auto oldOwnershipKind = SILValue(oldValue).getOwnershipKind();
-  if (oldOwnershipKind != OwnershipKind::None)
-    return true;
+  // 1. Compute the AccessPathWithBase of newValue. If we do not get back a
+  //    valid such object, invalidate and then bail.
+  //
+  // 2. Then we check if the base address is the result of an interior pointer
+  //    instruction. If we do not find one we bail.
+  //
+  // 3. Then grab the base value of the interior pointer operand. We only
+  //    support cases where we have a single BorrowedValue as our base. This is
+  //    a safe future proof assumption since one reborrows are on
+  //    structs/tuple/destructures, a guaranteed value will always be associated
+  //    with a single BorrowedValue, so this will never fail (and the code will
+  //    probably be DCEed).
+  //
+  // 4. Then we compute an AccessPathWithBase for oldValue and then find its
+  //    derived uses. If we fail, we bail.
+  //
+  // 5. At this point, we know that we can perform this RAUW. The only question
+  //    is if we need to when we RAUW copy the interior pointer base value. We
+  //    perform this check by making sure all of the old value's derived uses
+  //    are within our BorrowedValue's scope. If so, we clear the extra state we
+  //    were tracking (the interior pointer/oldValue's transitive uses), so we
+  //    perform just a normal RAUW (without inserting the copy) when we RAUW.
+  auto accessPathWithBase = AccessPathWithBase::compute(newValue);
```

Comment:
@atrick This is another case where I am nervous if I did this correctly (I just haven't played with the AccessPath APIs that much). One reason why I am unsure is due to the CSE test failure and some weird behavior I saw around this API and not stopping at pointer_to_address. I am going to dig in and try to understand what's going on.

Reply:
Based on the surrounding code, this looks like a correct and appropriate use of `AccessPathWithBase`. I haven't proven to my self yet that replaceAddress is complete yet.